src/http-client/index.ts

```diff
@@ -14,32 +33,112 @@ abstract class HttpClient extends Dispatcher {
    */
   public static async request(options: HttpClientRequestOptions): Promise<Response> {
     let res = await this._request(options, false);
-    if (this.newTokenRequired(res, options.requiresNewToken)) {
+    let txnAuthJSON: TxnAuthJSON | undefined;
+
+    if (newTokenRequired(res, options.requiresNewToken)) {
       res = await this._request(options, true);
     }
+
+    if (options.txnAuth && options.txnAuth.handleStep) {
+      if (res.redirected && examineForIGTxnAuth(res)) {
+        txnAuthJSON = normalizeIGJSON(res);
+      } else if (await examineForRESTTxnAuth(res)) {
+        txnAuthJSON = await normalizeRESTJSON(res);
+      }
+
+      if (txnAuthJSON && txnAuthJSON.advices) {
+        const { serverConfig } = Config.get(options.txnAuth.config);
+        const txnAuthOptions = buildTxnAuthOptions(
+          txnAuthJSON,
+          serverConfig.baseUrl,
+          options.timeout,
+        );
+        const initialStep = await this._request(txnAuthOptions, false);
+
+        if (await isAuthStep(initialStep)) {
+          try {
+            // Resolves with return of `void` upon successful completion
+            await this.stepIterator(initialStep, options.txnAuth.handleStep);
+            // Add Txn ID to original request options
+            options.txnAuth.txnID = txnAuthJSON.advices.TransactionConditionAdvice[0];
+            // Retry original resource request
+            res = await this._request(options, false);
+          } catch (err) {
+            throw new Error(err);
+          }
+        } else {
+          throw new Error('Error: Initial response from auth server not a "step".');
+        }
+      }
+    }
+
     return res;
   }
 
+  private static async stepIterator(res: Response, handleStep: HandleStep): Promise<void> {
+    const jsonRes = await res.json();
+    const initialStep = new FRStep(jsonRes);
+
+    return new Promise(async (resolve, reject) => {
+      async function handleNext(step: FRStep): Promise<void> {
+        const input = await handleStep(step);
+        const output = await FRAuth.next(input);
+
+        if (output.type === StepType.LoginSuccess) {
+          resolve();
+        } else if (output.type === StepType.LoginFailure) {
+          reject('Transaction authorization failure.');
+        } else {
+          handleNext(output);
+        }
+      }
+
+      handleNext(initialStep);
+    });
+  }
+
+  private static async setAuthHeaders(
```

Comment:
Setting headers got more and more complicated, so I pulled them out into their own method. In keeping with utility methods are pure stateless functions and private static functions handle side-effects, I made this a private static method.